src/test/descriptor_tests.cpp

```diff
@@ -42,33 +42,44 @@ bool EqualDescriptor(std::string a, std::string b)
     return a == b;
 }
 
-std::string MaybeUseHInsteadOfApostrophy(std::string ret)
+std::string UseHInsteadOfApostrophe(const std::string& desc)
 {
-    if (InsecureRandBool()) {
-        while (true) {
-            auto it = ret.find("'");
-            if (it != std::string::npos) {
-                ret[it] = 'h';
-                if (ret.size() > 9 && ret[ret.size() - 9] == '#') ret = ret.substr(0, ret.size() - 9); // Changing apostrophe to h breaks the checksum
-            } else {
-                break;
-            }
-        }
+    std::string ret = desc;
+    while (true) {
+        auto it = ret.find('\'');
+        if (it == std::string::npos) break;
+        ret[it] = 'h';
+    }
+
+    // GetDescriptorChecksum returns "" if the checksum exists but is bad.
+    // Switching apostrophes with 'h' breaks the checksum if it exists - recalculate it and replace the broken one.
+    if (GetDescriptorChecksum(ret) == "") {
+      ret = ret.substr(0, desc.size() - 9);
+      ret += std::string("#") + GetDescriptorChecksum(ret);
     }
     return ret;
 }
 
 const std::set<std::vector<uint32_t>> ONLY_EMPTY{{}};
 
-void Check(const std::string& prv, const std::string& pub, int flags, const std::vector<std::vector<std::string>>& scripts, const std::set<std::vector<uint32_t>>& paths = ONLY_EMPTY)
+void DoCheck(const std::string& prv, const std::string& pub, int flags, const std::vector<std::vector<std::string>>& scripts, const std::set<std::vector<uint32_t>>& paths = ONLY_EMPTY,
+    bool replace_apostrophe_with_h=false)
 {
     FlatSigningProvider keys_priv, keys_pub;
     std::set<std::vector<uint32_t>> left_paths = paths;
     std::string error;
 
+    std::unique_ptr<Descriptor> parse_priv;
+    std::unique_ptr<Descriptor> parse_pub;
     // Check that parsing succeeds.
-    auto parse_priv = Parse(MaybeUseHInsteadOfApostrophy(prv), keys_priv, error);
-    auto parse_pub = Parse(MaybeUseHInsteadOfApostrophy(pub), keys_pub, error);
+    if (replace_apostrophe_with_h) {
```

Comment:
There is actually a test case which used to be covered which is no longer ever covered - the use of `h` in the pub and `'` in the prv and vice versa (because there was a 50% chance of each being changed independently)

Reply:
Thanks for your review. You are right, I completely missed this. I added code to cover all four test cases (no replacement, prv only, pub only and prv+pub) in b9ee63c71b751fd67da777774ea8c0b27e7db2f8.